File: Nori/Panel/PanelSections.swift
import Foundation

/// Groups rows into PINNED / TODAY / YESTERDAY / EARLIER (or RESULTS while searching) and
/// assigns the positional ⌘1–9 numbers.
enum PanelSections {
    struct Section: Identifiable, Equatable, Sendable {
        var title: String
        var rows: [Row]
        var id: String { title }
    }

    struct Row: Identifiable, Equatable, Sendable {
        var row: ClipRow
        /// Highlight ranges in `row.title` for the current query.
        var titleRanges: [Range<String.Index>]
        /// ⌘1–9, positional over the visible list; nil past nine or for ghost rows.
        var number: Int?
        var id: UUID { row.id }
    }

    static func build(
        history: [ClipRow],
        sensitive: [ClipRow],
        ghosts: [ClipRow],
        filter: PanelFilter,
        query: String,
        now: Date = .now,
        calendar: Calendar = .current
    ) -> [Section] {
        let searching = !query.trimmingCharacters(in: .whitespaces).isEmpty

        var scored: [(Row, Double)] = []
        for row in history + sensitive where filter.matches(kind: row.kind) {
            if row.isSensitive, searching { continue }  // secrets are never indexed
            guard let match = HistorySearch.match(query: query, title: row.title, searchText: row.searchText) else { continue }
            // Fuzzy (subsequence) hits rank last and are not highlighted: scattered letters read as noise.
            let ranges = match.score >= 5 ? [] : match.titleRanges
            scored.append((Row(row: row, titleRanges: ranges, number: nil), match.score))
        }

        var sections: [Section] = []
        if searching {
            let ordered = scored.enumerated()
                .sorted { ($0.element.1, $0.offset) < ($1.element.1, $1.offset) }
                .map(\.element.0)
            if !ordered.isEmpty { sections.append(Section(title: String(localized: "Results"), rows: ordered)) }
        } else {
            let pinned = scored.map(\.0).filter { $0.row.isPinned }
                .sorted { ($0.row.pinnedAt ?? .distantPast) < ($1.row.pinnedAt ?? .distantPast) }
            let unpinned = scored.map(\.0).filter { !$0.row.isPinned }
                .sorted { $0.row.lastCopiedAt > $1.row.lastCopiedAt }

            if !pinned.isEmpty { sections.append(Section(title: String(localized: "Pinned"), rows: pinned)) }

            let ghostRows = filter == .all ? ghosts.map { Row(row: $0, titleRanges: [], number: nil) } : []
            var today: [Row] = ghostRows
            var yesterday: [Row] = []
            var earlier: [Row] = []
            for row in unpinned {
                if calendar.isDate(row.row.lastCopiedAt, inSameDayAs: now) {
                    today.append(row)
                } else if let dayBefore = calendar.date(byAdding: .day, value: -1, to: now),
                          calendar.isDate(row.row.lastCopiedAt, inSameDayAs: dayBefore) {
                    yesterday.append(row)
                } else {
                    earlier.append(row)
                }
            }
            if !today.isEmpty { sections.append(Section(title: String(localized: "Today"), rows: today)) }
            if !yesterday.isEmpty { sections.append(Section(title: String(localized: "Yesterday"), rows: yesterday)) }
            if !earlier.isEmpty { sections.append(Section(title: String(localized: "Earlier"), rows: earlier)) }
        }

        // Numbers are positional over the visible list; ghost rows are skipped.
        var number = 1
        for sectionIndex in sections.indices {
            for rowIndex in sections[sectionIndex].rows.indices where number <= 9 {
                guard !sections[sectionIndex].rows[rowIndex].row.isGhost else { continue }
                sections[sectionIndex].rows[rowIndex].number = number
                number += 1
            }
        }
        return sections
    }

    /// Relative time for the meta column: "now", "2m", "1h", "1d", "Sep 3".
    static func relativeTime(_ date: Date, now: Date = .now, calendar: Calendar = .current) -> String {
        let seconds = now.timeIntervalSince(date)
        if seconds < 60 { return String(localized: "now") }
        if seconds < 3_600 { let minutes = Int(seconds / 60); return String(localized: "\(minutes)m") }
        if seconds < 86_400 { let hours = Int(seconds / 3_600); return String(localized: "\(hours)h") }
        if seconds < 7 * 86_400 { let days = Int(seconds / 86_400); return String(localized: "\(days)d") }
        let formatter = DateFormatter()
        formatter.calendar = calendar
        formatter.setLocalizedDateFormatFromTemplate(calendar.isDate(date, equalTo: now, toGranularity: .year) ? "MMM d" : "MMM d yyyy")
        return formatter.string(from: date)
    }
}

File: windows/Nori.Core/PanelSections.cs
using System.Globalization;

namespace Nori.Core;

public enum SectionKind
{
    Pinned,
    Today,
    Yesterday,
    Earlier,
    Results,
}

/// <summary>
/// Groups rows into PINNED / TODAY / YESTERDAY / EARLIER (or RESULTS while searching) and
/// assigns the positional Ctrl+1–9 numbers.
/// </summary>
public static class PanelSections
{
    public sealed record Section(SectionKind Kind, IReadOnlyList<Row> Rows);

    /// <param name="Clip">The clip.</param>
    /// <param name="TitleRanges">Highlight ranges in <c>Row.Title</c> for the current query.</param>
    /// <param name="Number">Ctrl+1–9, positional over the visible list; null past nine or for ghost rows.</param>
    public sealed record Row(ClipRow Clip, IReadOnlyList<TextRange> TitleRanges, int? Number)
    {
        public Guid Id => Clip.Id;
    }

    public static IReadOnlyList<Section> Build(
        IReadOnlyList<ClipRow> history,
        IReadOnlyList<ClipRow> sensitive,
        IReadOnlyList<ClipRow> ghosts,
        PanelFilter filter,
        string query,
        DateTimeOffset now,
        TimeZoneInfo? timeZone = null)
    {
        timeZone ??= TimeZoneInfo.Local;
        var searching = query.Trim().Length > 0;

        var scored = new List<(Row Row, double Score)>();
        foreach (var row in history.Concat(sensitive))
        {
            if (!filter.Matches(row.Kind)) continue;
            if (row.IsSensitive && searching) continue; // secrets are never indexed
            var match = HistorySearch.Find(query, row.Title, row.SearchText);
            if (match is null) continue;
            scored.Add((new Row(row, match.TitleRanges, null), match.Score));
        }

        var sections = new List<(SectionKind Kind, List<Row> Rows)>();
        if (searching)
        {
            var ordered = scored
                .Select((entry, index) => (entry, index))
                .OrderBy(x => x.entry.Score)
                .ThenBy(x => x.index)
                .Select(x => x.entry.Row)
                .ToList();
            if (ordered.Count > 0) sections.Add((SectionKind.Results, ordered));
        }
        else
        {
            var pinned = scored.Select(s => s.Row).Where(r => r.Clip.IsPinned)
                .OrderBy(r => r.Clip.PinnedAt ?? DateTimeOffset.MinValue).ToList();
            var unpinned = scored.Select(s => s.Row).Where(r => !r.Clip.IsPinned)
                .OrderByDescending(r => r.Clip.LastCopiedAt).ToList();

            if (pinned.Count > 0) sections.Add((SectionKind.Pinned, pinned));

            var today = filter == PanelFilter.All ? ghosts.Select(g => new Row(g, [], null)).ToList() : [];
            var yesterday = new List<Row>();
            var earlier = new List<Row>();
            var todayDate = LocalDate(now, timeZone);
            foreach (var row in unpinned)
            {
                var day = LocalDate(row.Clip.LastCopiedAt, timeZone);
                if (day == todayDate) today.Add(row);
                else if (day == todayDate.AddDays(-1)) yesterday.Add(row);
                else earlier.Add(row);
            }
            if (today.Count > 0) sections.Add((SectionKind.Today, today));
            if (yesterday.Count > 0) sections.Add((SectionKind.Yesterday, yesterday));
            if (earlier.Count > 0) sections.Add((SectionKind.Earlier, earlier));
        }

        // Numbers are positional over the visible list; ghost rows are skipped.
        var number = 1;
        var result = new List<Section>();
        foreach (var (kind, rows) in sections)
        {
            var numbered = new List<Row>(rows.Count);
            foreach (var row in rows)
            {
                if (!row.Clip.IsGhost && number <= 9)
                {
                    numbered.Add(row with { Number = number });
                    number++;
                }
                else
                {
                    numbered.Add(row);
                }
            }
            result.Add(new Section(kind, numbered));
        }
        return result;
    }

    public static DateOnly LocalDate(DateTimeOffset moment, TimeZoneInfo timeZone) =>
        DateOnly.FromDateTime(TimeZoneInfo.ConvertTime(moment, timeZone).DateTime);

    /// <summary>Relative time for the meta column: "now", "2m", "1h", "1d", "Sep 3".</summary>
    public static string RelativeTime(DateTimeOffset date, DateTimeOffset now, CultureInfo? culture = null, TimeZoneInfo? timeZone = null)
    {
        culture ??= CultureInfo.InvariantCulture;
        timeZone ??= TimeZoneInfo.Local;
        var seconds = (now - date).TotalSeconds;
        if (seconds < 60) return "now";
        if (seconds < 3_600) return $"{(int)(seconds / 60)}m";
        if (seconds < 86_400) return $"{(int)(seconds / 3_600)}h";
        if (seconds < 7 * 86_400) return $"{(int)(seconds / 86_400)}d";
        var local = TimeZoneInfo.ConvertTime(date, timeZone);
        var localNow = TimeZoneInfo.ConvertTime(now, timeZone);
        var sameYear = local.Year == localNow.Year;
        var japanese = culture.TwoLetterISOLanguageName == "ja";
        var format = (japanese, sameYear) switch
        {
            (true, true) => "M月d日",
            (true, false) => "yyyy年M月d日",
            (false, true) => "MMM d",
            (false, false) => "MMM d yyyy",
        };
        return local.ToString(format, culture);
    }

    /// <summary>"Sep 3, 14:02" for the expanded meta strip.</summary>
    public static string AbsoluteTime(DateTimeOffset date, CultureInfo? culture = null, TimeZoneInfo? timeZone = null)
    {
        culture ??= CultureInfo.InvariantCulture;
        timeZone ??= TimeZoneInfo.Local;
        var local = TimeZoneInfo.ConvertTime(date, timeZone);
        var format = culture.TwoLetterISOLanguageName == "ja" ? "M月d日 HH:mm" : "MMM d, HH:mm";
        return local.ToString(format, culture);
    }
}
